// @sei-js/cosmos/dist/esm/rest/evm/query.js
import * as fm from "../fetch";
export class Query {
    static SeiAddressByEVMAddress(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/sei_address?${fm.renderURLSearchParams(req, [])}`, { ...initReq, method: "GET" });
    }
    static EVMAddressBySeiAddress(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/evm_address?${fm.renderURLSearchParams(req, [])}`, { ...initReq, method: "GET" });
    }
    static StaticCall(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/static_call?${fm.renderURLSearchParams(req, [])}`, {
            ...initReq,
            method: "GET"
        });
    }
    static Pointer(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/pointer?${fm.renderURLSearchParams(req, [])}`, {
            ...initReq,
            method: "GET"
        });
    }
    static PointerVersion(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/pointer_version?${fm.renderURLSearchParams(req, [])}`, { ...initReq, method: "GET" });
    }
    static Pointee(req, initReq) {
        return fm.fetchReq(`/sei-protocol/seichain/evm/pointee?${fm.renderURLSearchParams(req, [])}`, {
            ...initReq,
            method: "GET"
        });
    }
}
