// @sei-js/confidential-transfers/dist/cjs/payload/confidentialApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfidentialTransfersWrapper = void 0;
const confidentialTransfers_1 = require("./confidentialTransfers");
require("./wasm_exec.cjs");
class ConfidentialTransfersWrapper {
    constructor() {
        this.wasmWrapper = new confidentialTransfers_1.ConfidentialTransfers();
        this.wasmWrapper.initialize();
    }
    async initialize() {
        await this.wasmWrapper.initialize();
    }
    createWithdraw(signedDenom, amount, address, denom, currentDecryptableBalance, currentAvailableBalance) {
        const result = this.wasmWrapper.createWithdraw(signedDenom, amount, address, denom, currentDecryptableBalance, currentAvailableBalance);
        return result;
    }
    createInitialize(signedDenom, address, denom) {
        const result = this.wasmWrapper.createInitialize(signedDenom, address, denom);
        return result;
    }
    createApplyPendingBalance(signedDenom, address, denom, pendingBalanceCreditCounter, currentDecryptableBalance, currentAvailableBalance, currentPendingBalanceLo, currentPendingBalanceHi) {
        const result = this.wasmWrapper.createApplyPendingBalance(signedDenom, address, denom, pendingBalanceCreditCounter, currentDecryptableBalance, currentAvailableBalance, currentPendingBalanceLo, currentPendingBalanceHi);
        return result;
    }
    createTransfer(signedDenom, amount, senderAddress, recipientAddress, denom, currentDecryptableBalance, currentAvailableBalance, recipientPubkey) {
        const result = this.wasmWrapper.createTransfer(signedDenom, amount, senderAddress, recipientAddress, denom, currentDecryptableBalance, currentAvailableBalance, recipientPubkey);
        return result;
    }
    createCloseAccount(signedDenom, address, denom, currentAvailableBalance, currentPendingBalanceLo, currentPendingBalanceHi) {
        const result = this.wasmWrapper.createCloseAccount(signedDenom, address, denom, currentAvailableBalance, currentPendingBalanceLo, currentPendingBalanceHi);
        return result;
    }
    decryptCiphertext(signedDenom, ciphertext) {
        const result = this.wasmWrapper.decryptCiphertext(signedDenom, ciphertext);
        return result;
    }
    decryptAesCiphertext(signedDenom, aesCiphertext) {
        const result = this.wasmWrapper.decryptAESCiphertext(signedDenom, aesCiphertext);
        return result;
    }
}
exports.ConfidentialTransfersWrapper = ConfidentialTransfersWrapper;
